### PR TITLE
Faster compiles

### PR DESCRIPTION
**Pull request by evincarofautumn**
WIP/RFC. This attempts to speed up compilations in a couple of ways:
- It introduces forward-declaration headers to avoid reincluding large headers, while also avoiding redeclaration of typedefs, a C11 feature. Doing this forces more source files to include what they use.
- It builds the subdirectories of `mono` concurrently, and makes the dependencies between those subdirectories explicit.

### CODE REVIEW

File: mono/utils/monobitset.h [hunk 1]
```diff
@@ -9,26 +9,15 @@
 #include <mono/utils/mono-publib.h>
 #endif
 
-/*
```
Comment (tritao) [inline mono/utils/monobitset.h]: I think this change is big enough that it should be in a different commit. Also why add it both to publib and compiler headers?

**Comment by tritao** [comment]
Looks good to me.

**Comment by alexrp** [comment]
Any benchmarks?

File: mono/utils/monobitset.h (hunk 1 above)
Comment (evincarofautumn) [inline mono/utils/monobitset.h]: There was somewhere that included one or the other based on an `#ifdef`. I will check it out.

**Comment by kumpera** [comment]
This requires C11, does all our targets support it?

**Comment by alexrp** [comment]
MSVC doesn't.

**Comment by evincarofautumn** [comment]
What requires C11? I meant that I was trying to avoid redeclaration of typedefs *because* it is a C11 feature. That’s why I placed them in headers with include guards, rather than inline.

**Comment by kumpera** [comment]
Err, I misread you commit message.

**Comment by kumpera** [comment]
Jon, how would you describe the mechanics of what you did?

This to remain useful we must enforce it as we move forward, so have it properly described would help others following it.

**Comment by evincarofautumn** [comment]
It’s still just an experiment. Merely using ccache produces much more dramatic improvements than any of these changes so far, at least for clean builds. Here is my process:

 * Use [this script](https://gist.github.com/evincarofautumn/88a3bd05e0c5cf102796) to find large headers that are included frequently by other headers, where the compiler may be doing redundant work.
 * For a header `x.h`, add a header `x-forward.h`.
 * Move all `struct _X;` and `typedef struct _X X;` declarations to `x-forward.h`.
 * Include `x-forward.h` from `x.h`.
 * For any header `y.h` that includes `x.h`, see if `y.h` can include only `x-forward.h`. If so, change it and fix anything that depended on `y.h` including `x.h` directly.

**Comment by vargaz** [comment]
These forwarded types make 'go to definition' etc. in editors harder to use, since using it on MonoObject for example, will go to the
typedef struct _MonoObject MonoObject definition, which is not very useful.

**Comment by evincarofautumn** [comment]
I’m not sure what you mean. These typedefs already exist, they’re just being moved. They’re usually separate from the `struct` definitions, anyway; `MonoObject` is an exception. With ctags, I already have the problem you describe; if I `find-tag` on `MonoMethod`, it takes me to the `typedef`, then I have to `find-tag` on `_MonoMethod`.

**Comment by vargaz** [comment]
Then its probably not going to make things worse.

**Comment by lewurm** [comment]
is there still interest to merge this or should it be closed?

**Comment by schani** [comment]
It would be nice if we could run this on Touchstone, if rebasing it isn't too hard.
